README.md



### Run in Debug Mode
DEBUG=scannerApp:* npm run devstart

### Build Docker Container
docker build -t local/scanner-app . 

### Run Docker Container
docker run -p 3000:3000 local/scanner-app

api/CHANGELOG.txt
ver-001
: Basic app (successfully hosted on Beanstalk)

ver-002
: Added /returndata endpoint for returing all data
: Changed the successful negative responses from 4xx to 2xx
: Now all calls return "statuscode" and "description"
: /authenticate returns "first", "last" and "email" instead of "id"

ver-003
: Removed JSON file storage
: Replaced storage with MySQL database
: Removed field "id" from database



app.js
var createError = require('http-errors');
var express = require('express');
var path = require('path');
var cookieParser = require('cookie-parser');
var logger = require('morgan');

var indexRouter = require('./routes/index');
var enrollRouter = require('./routes/enroll');
var authenticateRouter = require('./routes/authenticate');

var app = express();

// view engine setup
app.set('views', path.join(__dirname, 'views'));
app.set('view engine', 'ejs');

app.use(logger('dev'));
app.use(express.json());
app.use(express.urlencoded({ extended: false }));
app.use(cookieParser());
app.use(express.static(path.join(__dirname, 'public')));

app.use('/', indexRouter);
app.use('/enroll', enrollRouter);
app.use('/authenticate', authenticateRouter);

// catch 404 and forward to error handler
app.use(function(req, res, next) {
  next(createError(404));
});

// error handler
app.use(function(err, req, res, next) {
  // set locals, only providing error in development
  res.locals.message = err.message;
  res.locals.error = req.app.get('env') === 'development' ? err : {};

  // render the error page
  res.status(err.status || 500);
  res.render('error');
});

module.exports = app;


api/utils.py
import os
import re
import mysql.connector

def get_database():
    mydb = mysql.connector.connect(
        host="fingervein-api-data.c7agb2tyg4gq.us-east-1.rds.amazonaws.com",
        user="admin",
        password="emtechiot",
        database="fingervein"
    )
    return mydb

def get_user_from_finger(val):
    mydb = get_database()
    cur = mydb.cursor()
    sqlquery = "SELECT first, last, email FROM apidata WHERE fingertemplate = %s"
    sqldata = (val, )
    cur.execute(sqlquery, sqldata)
    res = cur.fetchall()
    mydb.close()
    if res == []:
        return -1, {}

    data = {"first" : res[0][0],
            "last" : res[0][1],
            "email" : res[0][2]}
    return 0, data

def check_email_exist(val):
    mydb = get_database()
    cur = mydb.cursor()
    sqlquery = "SELECT COUNT(*) FROM apidata WHERE email = %s"
    sqldata = (val, )
    cur.execute(sqlquery, sqldata)
    res = cur.fetchall()
    mydb.close()
    if res[0][0]:
        return 1
    return 0

def add_datapoint(newval):
    mydb = get_database()
    cur = mydb.cursor()
    sqlquery = "SELECT COUNT(*) FROM apidata WHERE email = %s"
    sqldata = (newval["email"], )
    cur.execute(sqlquery, sqldata)
    res = cur.fetchall()
    if res[0][0]:
        mydb.close()
        return 0
    
    sqlquery = "INSERT INTO apidata (first, last, email, fingertemplate, fingerverificationtemplate) VALUES (%s, %s, %s, %s, %s)"
    sqldata = (newval["first"], newval["last"], newval["email"], newval["fingertemplate"], newval["fingerverificationtemplate"])
    cur.execute(sqlquery, sqldata)
    mydb.commit()
    mydb.close()
    return 1

def delete_email(val):
    mydb = get_database()
    cur = mydb.cursor()
    sqlquery = "DELETE FROM apidata WHERE email = %s"
    sqldata = (val, )
    cur.execute(sqlquery, sqldata)
    mydb.commit()
    mydb.close()
    return cur.rowcount

def reset_database():
    mydb = get_database()
    cur = mydb.cursor()
    sqlquery = "DELETE FROM apidata"
    cur.execute(sqlquery)
    mydb.commit()
    mydb.close()
    return cur.rowcount

def get_alldata():
    mydb = get_database()
    cur = mydb.cursor()
    sqlquery = "SELECT * FROM apidata"
    cur.execute(sqlquery)
    a = cur.fetchall()
    mydb.close()

    alldata = []
    for obj in a:
        curdata = {
            "first" : obj[0],
            "last" : obj[1],
            "email" : obj[2],
            "fingertemplate" : obj[3],
            "fingerverificationtemplate" : obj[4]
        }
        alldata.append(curdata)
    return alldata

public/javascripts/enrollment.js

var h1Plugin = new pluginManager('pluginDiv');
var currentTab = 0; // Current tab is set to be the first tab (0)

var cors_api_url = "https://cors-anywhere.herokuapp.com/"
var url = cors_api_url + "http://fingerveinrestapi-env.eba-yv7gqaek.us-east-1.elasticbeanstalk.com/"

$(document).ready(function (e){
    h1Plugin.pageLoaded();

    $(".alert").hide();

    showTab(currentTab); // Display the current tab
   
    $("#accept-checkbox").click(function() {
        if(this.checked) {
          toggleButtonState('#nextBtn', false);
        }
        else {
          toggleButtonState('#nextBtn', true);
        }
    });

});

function showTab(n) {
  // This function will display the specified tab of the form ...
  var x = $(".tab")
  x[n].style.display = "block";

  // ... and fix the Previous/Next buttons:
  if (n == 0) {
    $("#prevBtn").css("display", "none");
  } else {
    $("#prevBtn").css("display", "inline");
  }

  if (n == (x.length - 1)) {
    $("#nextBtn").innerHTML = "Submit";
  } else {
   $("#nextBtn").innerHTML = "Next";
  }

  // ... and run a function that displays the correct step indicator:
  fixStepIndicator(n)
}

function nextPrev(n) {

  // Disable Next button until terms have been accepted
  if( currentTab == 0) {
    toggleButtonState('#nextBtn', true);
  }

  if(currentTab == 2) {
    // Verify user is not already enrolled
    checkUserExits(n);
  }
  else if( currentTab == 4 ) {
    // Start enrollment process
    toggleButtonState('#nextBtn', true);
    h1Plugin.enrollCapturedTemplate( captureTemplateSuccess, scanError );
    console.log("Enrolling")
  }
  else {
    showNextPrev(n);
  }
    
}

function showNextPrev(n) {
  // This function will figure out which tab to display
  var x = $(".tab")

  // Exit the function if any field in the current tab is invalid:
  if (n == 1 && !validateForm()) return false;

  // Hide the current tab:
  x[currentTab].style.display = "none";

  // Increase or decrease the current tab by 1:
  currentTab = currentTab + n;

  // if you have reached the end of the form... :
  if (currentTab >= x.length) {
    //...the form gets submitted:
    document.getElementById("enrollForm").submit();
    return false;
  }

  // Otherwise, display the correct tab:
  showTab(currentTab);
}

function checkUserExits(n) {
  $.ajax({
    url: url + "checkexists",
    type: "POST",
    contentType: "application/json",
    dataType: "json",
    data: JSON.stringify({
      email: $('#email').val()
    }),
    success: function(data) {
      console.log(data);

      if(data.statuscode == -1) {
        showNextPrev(n);
      }
      else {
        displayAlert("User Already Exists.");
      }
    },
    error: function(error) {
      console.log(error);
      displayAlert("User Already Exists.");
    }
  });
}

function enrollUserData() {
  $.ajax({
    url: url + "enroll",
    type: "POST",
    contentType: "application/json",
    dataType: "json",
    data: JSON.stringify({
      first: $('#first').val(),
      last: $('#last').val(),
      email: $('#email').val(),
      fingertemplate: $('#hdnFinger1').val(),
      fingerverificationtemplate: $('#hdnVerifyFinger1').val()
    }),
    success: function(data) {
      console.log(data);

      if(success == 200) {
        showNextPrev(1);
      }
    },
    error: function(error) {
      console.log(error);
      showNextPrev(1);
    }
  });
}

function captureTemplateSuccess(enrollTemplate) {

  if (enrollTemplate.indexOf("Warning") != -1) {
    $("#alert-text").text(enrollTemplate);
    return false;
  }

  console.log("Enroll Template:")
  console.log(enrollTemplate)
  $("#hdnFinger1").val(enrollTemplate);

  h1Plugin.verifyEnrollment( verifyTemplateSuccess, scanError );
}

function verifyTemplateSuccess(verifyAfterTemplate) {
  if (verifyAfterTemplate.indexOf("Warning") != -1) {
    alertError(verifyAfterTemplate);
    return false;
  }

  console.log("Verify Template:")
  console.log(verifyAfterTemplate);
  $("#hdnVerifyFinger1").val(verifyAfterTemplate);
  enrollUserData();
}

function scanError(errorInfo, option) {
  console.log(option + ": " + errorInfo);
  $("#alert-text").text(option + ": " + errorInfo);
}

function validateForm() {
  // This function deals with validation of the form fields
  var x, y, i, valid = true;
  x = document.getElementsByClassName("tab");
  y = x[currentTab].getElementsByTagName("input");
  // A loop that checks every input field in the current tab:
  for (i = 0; i < y.length; i++) {
    // If a field is empty...
    if (y[i].value == "") {
      // add an "invalid" class to the field:
      y[i].className += " invalid";
      // and set the current valid status to false:
      valid = false;
    }
  }
  // If the valid status is true, mark the step as finished and valid:
  if (valid) {
    document.getElementsByClassName("step")[currentTab].className += " finish";
  }
  return valid; // return the valid status
}

function fixStepIndicator(n) {
  // This function removes the "active" class of all steps...
  var i, x = document.getElementsByClassName("step");
  for (i = 0; i < x.length; i++) {
    x[i].className = x[i].className.replace(" active", "");
  }
  //... and adds the "active" class to the current step:
  x[n].className += " active";
}

function toggleButtonState(button, disabled) {
  if(disabled){
    $(button).attr('disabled', true); //disable input
    $(button).css('background-color', "#bbbbbb"); //disable input
    $(button).css('cursor', "pointer"); //disable input
  }
  else {
    $(button).attr('disabled', false); //disable input
    $(button).css('background-color', "#4CAF50"); //disable input
    $(button).css('cursor', "pointer"); //disable input
  }
}

function displayAlert(msg) {
  $("#alert-text").text(msg);
  $(".alert").show();
}




public/javascripts/authentication.js
var h1Plugin = new pluginManager('pluginDiv');
var currentTab = 0; // Current tab is set to be the first tab (0)

var cors_api_url = "https://cors-anywhere.herokuapp.com/"
var url = cors_api_url + "http://fingerveinrestapi-env.eba-yv7gqaek.us-east-1.elasticbeanstalk.com/"

$(document).ready(function (e){
    h1Plugin.pageLoaded();

    $(".alert").hide();
    $("#prevBtn").css("display", "none");

    showTab(currentTab); // Display the current tab
});

function showTab(n) {
    // This function will display the specified tab of the form ...
    var x = $(".tab")
    x[n].style.display = "block";
  
    if (n == (x.length - 1)) {
      $("#nextBtn").innerHTML = "Submit";
    } else {
     $("#nextBtn").innerHTML = "Proceed";
    }
  
    // ... and run a function that displays the correct step indicator:
    fixStepIndicator(n)
}

function nextPrev(n) {
    // This function will figure out which tab to display
    var x = $(".tab")

    if(currentTab == 0) {
        // Start enrollment process
        h1Plugin.authentication( authenticateSuccess, authenticatError );
        console.log("Authenticating")
    }
    else {
        // Hide the current tab:
        x[currentTab].style.display = "none";

        // Increase or decrease the current tab by 1:
        currentTab = currentTab + n;

        // if you have reached the end of the form... :
        if (currentTab >= x.length) {
        //...the form gets submitted:
        document.getElementById("enrollForm").submit();
        return false;
        }

        // Otherwise, display the correct tab:
        showTab(currentTab); 
    }
}

function showNextPrev(n) {
    // This function will figure out which tab to display
    var x = $(".tab")
  
    // Hide the current tab:
    x[currentTab].style.display = "none";
  
    // Increase or decrease the current tab by 1:
    currentTab = currentTab + n;
  
    // if you have reached the end of the form... :
    if (currentTab >= x.length) {
      //...the form gets submitted:
      document.getElementById("enrollForm").submit();
      return false;
    }
  
    // Otherwise, display the correct tab:
    showTab(currentTab);
}

function authenticateSuccess(enrollTemplate) {

    console.log("Authenticate Template:")
    console.log(enrollTemplate);

    if ((enrollTemplate.indexOf("Warning") != -1) || (enrollTemplate.indexOf("Error") != -1)) {
        displayAlert(enrollTemplate);
        return false;
    }

    if (enrollTemplate != null && enrollTemplate != "")
    {
        $.ajax({
            url: url + "authenticate",
            type: "POST",
            contentType: "application/json",
            dataType: "json",
            data: JSON.stringify({
              fingertemplate: enrollTemplate,
            }),
            success: function(data) {
                console.log(data);
        
                // Set Data and show Results screen
                showNextPrev(1);
            },
            error: function(error) {
              console.log(error);
              showNextPrev(1);
            }
        });
    }
}

function authenticatError(errorInfo) {
    console.log("Error")
}

function fixStepIndicator(n) {
    // This function removes the "active" class of all steps...
    var i, x = document.getElementsByClassName("step");
    for (i = 0; i < x.length; i++) {
      x[i].className = x[i].className.replace(" active", "");
    }
    //... and adds the "active" class to the current step:
    x[n].className += " active";
}

function displayAlert(msg) {
    $("#alert-text").text(msg);
    $(".alert").show();
  }
  

public/javascripts/plugin_manager.js
var pluginManager = function() {
  var mimetype = "application/x-fviescannerplugin"
  var plugin = null;

  this.pluginLoaded = function() {
		plugin = this.getPlugin();
  };
  
  this.pageLoaded = function(){

    var parser = new UAParser();
    var browser = parser.getBrowser().name
    var browserVersion = parser.getBrowser().version

		switch(browser)
		{
			case "Chrome":
				if(browserVersion < 62)
					alert("The scanner plug-in requires Chrome version:62 or above.");
				else
					initPlugin();
				break;
			case "IE":
        initializeActiveXOrNPAPIPlugin();
				break;
			case "Firefox":
				if(browserVersion < 52)
					initializeActiveXOrNPAPIPlugin();
				else
					initPlugin();

				break;
		}
  };

  this.getPlugin = function () {
		if(plugin != null)
		{
			return plugin;
		}
		else
		{
			plugin = document.getElementById("h1FvPlugin");
			return plugin;
		}
  };
  
  this.enrollCapturedTemplate = function ( successCallback, errorCallback ) {

    var finger = "1", timeout = 20000, showPrompt = true;
    var pluginInstance = this.getPlugin();

		pluginInstance.getEnrolmentTemplate(finger, timeout, showPrompt).then(
		
		  function(fvTemplate) {

				if(isSuccess(fvTemplate)) {
          if(successCallback)
						successCallback(fvTemplate);
				}
				else {
          if(errorCallback)
						errorCallback(fvTemplate, "Capture");
				}
			},
			
			function(errorInfo) {
        if(errorCallback)
          errorCallback(errorInfo, "Capture");
			}
		);
  }

  this.verifyEnrollment = function( successCallback, errorCallback ) {
    
    var finger = "1", timeout = 20000, threshold= 1000, showPrompt = true;
    var pluginInstance = this.getPlugin();

		pluginInstance.verifyAfterEnrol(finger, timeout, threshold, showPrompt).then(
		
      function(fvTemplate) {
        
        if(isSuccess(fvTemplate))
				{
					if(successCallback)
						successCallback(fvTemplate);
				}
				else
				{
					if(errorCallback)
						errorCallback(fvTemplate, "Verify");
				}
			},
			
			function(errorInfo) {
        if(errorCallback)
          errorCallback(errorInfo, "Verify");
			}
		);
  }

  this.authentication = function( successCallback, errorCallback ) {
    var finger = "1", timeout = 20000, showPrompt = true;
    var pluginInstance = this.getPlugin();

		pluginInstance.getVerificationTemplate(finger, timeout, showPrompt).then(
		
      function(fvTemplate) {
        if(isSuccess(fvTemplate))
				{
					if(successCallback)
						successCallback(fvTemplate);
				}
				else
				{
					if(errorCallback)
						errorCallback(errorCallback);
				}
      },
			function(errorInfo){
        errorCallback(errorInfo);
			}
		);

  }

  function initPlugin() {
    //Create a var that basically saves a random number to guarantee an unique identifier for a function. _plugin_ can be any other name.
    var callbackFn = "_plugin_" + Math.floor(Math.random() * 100000000);
  
    window[callbackFn] = function (data) {
      //Retrieve the wyrmhole factory for later creation
      var helper = data.wyrmhole;
  
      setHelper(helper);
    }
    /*Post a message to the extension, telling it to instantiate a a wyrmhole.
    FBDevTeam should be the name of your company inside the plugin configuration. For the echoTestPlugin its FBDevTeam.
    callbackFn is the function that will be called once the result of the postMessage is returned.
    */
    window.postMessage({firebreath: 'hitachi', callback: callbackFn}, "*");
  }

  function initializeActiveXOrNPAPIPlugin() {
		var script = document.createElement("script");
		script.type = "text/javascript";
		script.innerHTML = "function pluginLoaded() { h1Plugin.pluginLoaded(); }";
		document.head.appendChild(script);
		document.getElementById("pluginDiv").innerHTML = pluginTag;
	}
  
  function setHelper(helper) {
    //Using the wyrmholeFactory we create a wyrmhole.
    helper.create(mimetype).then(
        function (wyrmhole) {
          //With the created wyrmhole we instantiate a new FireWyrmJS object that will allow us to create the plugin.
          var FWJS = window.FireWyrmJS;
          //Create pluginFactory that will allow the plugin creation.
          window.pluginFactory = new FWJS(wyrmhole);
        pluginFactory.create(mimetype, {/*some params*/}).then(
          function (pluginObj) {
            //Save the plugin to a gloal var for later access
            plugin = pluginObj;
            console.log("Wyrmhole created");
          },
          function (error) {
            console.log("An Unexpected Error has ocurred: ", error);
          })
        },
        function (error) {
          console.log("An Unexpected Error has ocurred: ", error);
        }
      )
  }

  function getPluginVersion() {
    var pluginInstance = this.getPlugin();
		pluginInstance.version.then(
			function(versionInfo) {
				console.log("Plugin Version Is: ", versionInfo);
			},
			function(errorInfo) {
				console.log("An Unexpected Error has ocurred: ", errorInfo);
			}
		);
  }

  function getAppName() {
    var pluginInstance = this.getPlugin();
    pluginInstance.getAppName().then(
      function(appName) {
        console.log("Application Name: ", appName);
      },
        
      function(errorInfo){
        console.log("An Unexpected Error has ocurred: ", errorInfo);
      }
		);
  }
  
  function getUserSID() {
    var pluginInstance = this.getPlugin();
		pluginInstance.getUserSID().then(
      function(userSID){
        console.log("User ID: ", userSID);
      },
			
			function(errorInfo) {
				console.log("An Unexpected Error has ocurred: ", errorInfo);
			}
		);
  }

  var isSuccess = function(resultStr)
	{
		var isSuccessResponse = true;
		var response = resultStr.trim();
		
		if(!response || response.length === 0)
			isSuccessResponse = false;
		else if(response.toUpperCase().lastIndexOf("Error".toUpperCase()) === 0)
			isSuccessResponse = false;
		else if(response.toUpperCase().lastIndexOf("Warning".toUpperCase()) === 0)
			isSuccessResponse = false;
		else
			isSuccessResponse = true;
		
		return isSuccessResponse;
  };
  
};





  

api/application.py
from flask import Flask, jsonify, request, abort
import json
from utils import add_datapoint, get_user_from_finger, check_email_exist, delete_email, reset_database, get_alldata

application = Flask(__name__)

@application.route('/')
def homepage():
    return "RESTFUL API for FingerVein Running ver-003"

@application.route('/enroll', methods=['POST'])
def enroll():
    if not request.json or "first" not in request.json \
                        or "last" not in request.json \
                        or "email" not in request.json \
                        or "fingertemplate" not in request.json \
                        or "fingerverificationtemplate" not in request.json:
        return jsonify({"statuscode" : 400,
                        "description" : "No / Incomplete data given for enrollment"}), 400
    
    new_user = {
        "first" : request.json["first"],
        "last" : request.json["last"],
        "email" : request.json["email"],
        "fingertemplate" : request.json["fingertemplate"],
        "fingerverificationtemplate" : request.json["fingerverificationtemplate"]
    }
    if not add_datapoint(new_user):
        return jsonify({"statuscode" : -1,
                        "description" : "User already exists"}), 200
    return jsonify({"statuscode" : 0,
                    "description" : "User successfully added"}), 200

@application.route('/authenticate', methods=['POST'])
def authentication():
    if not request.json or "fingertemplate" not in request.json:
        return jsonify({"statuscode" : 400,
                        "description" : "No finger template provided"}), 400
    
    fingercode = request.json["fingertemplate"]
    response, responsedata = get_user_from_finger(fingercode)
    if response == -1:
        return jsonify({"statuscode" : -1,
                        "description" : "Finger template not authenticated"}), 200
    return jsonify({"statuscode" : 0,
                    "description" : "User authenticated",
                    "data" : responsedata}), 200

@application.route('/checkexists', methods=['POST'])
def checkexists():
    if not request.json or "email" not in request.json:
        return jsonify({"statuscode" : 400,
                        "description" : "No email provided"}), 400

    email = request.json["email"]
    response = check_email_exist(email)
    if response:
        return jsonify({"statuscode" : 0,
                        "description" : "User with this email exists"}), 200
    return jsonify({"statuscode" : -1,
                    "description" : "User with this email does not exist"}), 200

@application.route('/delete', methods=['POST'])
def delete():
    if not request.json or "email" not in request.json:
        return jsonify({"statuscode" : 400,
                        "description" : "No email provided"}), 400

    email = request.json["email"]
    response = delete_email(email)
    if response:
        return jsonify({"statuscode" : 0,
                        "description" : str(response) + " record(s) deleted"}), 200
    return jsonify({"statuscode" : -1,
                    "description" : "No entry found with the given email"}), 200

@application.route('/reset', methods=['GET'])
def reset():
    response = reset_database()
    return jsonify({"statuscode" : 0,
                    "description" : str(response) + " (ALL) record(s) deleted"}), 200

@application.route('/returndata', methods=['GET'])
def returndata():
    data = get_alldata()
    return jsonify({"statuscode" : 0,
                    "description" : "All data returned",
                    "data" : data}), 200

if __name__ == '__main__':
    application.run(host="0.0.0.0", port=80)